### ISSUE TITLE
Make `T::Struct` use same equality semantics as Ruby core `Struct`

### ISSUE THREAD

**Issue by fajpunk** [sorbet/sorbet#1540]
### Problem

`T::Struct` does not have the same equality semantics as core Ruby `Struct`s.  Ruby core `Struct`s define `==` as:

```
Returns true if other has the same struct subclass and has equal member values (according to Object#==).
```

Whereas `T::Struct`s are using `Object`'s `==` (I think).

I would expect that `T::Struct`s use the same equality semantics as Ruby core `Struct`s.

```ruby
#!/usr/bin/env ruby
require 'sorbet-runtime'

class MyTypedStruct < T::Struct
  prop :my_prop, Integer
end

ts1 = MyTypedStruct.new(my_prop: 1)
ts2 = MyTypedStruct.new(my_prop: 1)

if ts1 == ts2
  puts "T::Struct's are equal"
else
  puts "T::Struct's are NOT equal"
end

MyStruct = Struct.new(:my_prop)

s1 = MyStruct.new(1)
s2 = MyStruct.new(2)

if s1 == s2
  puts "Struct's are equal"
else
  puts "Struct's are NOT equal"
end
```
```
T::Struct's are equal
Struct's are equal
```

### Proposed solution

If others agree that `T::Struct` equality should be like Ruby core `Struct` equality, then I'm not sure offhand what the best way to implement it would be.  Ruby's `Struct#==` is implemented in C.  I'm not sure what the most efficient way would be to implement it for `T::Struct`.

**Comment by jez** [comment]
I don't think we're super opposed to this change. The biggest problem with implementing it is just that we have thousands of existing `T::Struct`s in use at Stripe, and it's very possible that some of them depend on the current behavior. If someone were to implement this, we couldn't land the change until we were sure that it doesn't break any existing usages at Stripe, or if it does break existing usages that they're first fixed.

**Comment by jez** [comment]
After looking into this a little out of curiosity, it looks like Ruby's `Struct#==` implements some really gnarly custom logic to make it such that

```ruby
Box = Struct.new(:val)

b1 = Box.new(nil)
b2 = Box.new(nil)

b1.val = b2
b2.val = b1

puts b1 == b2
```

doesn't infinitely loop and in fact returns `true`.

=> https://github.com/ruby/ruby/blob/master/thread.c#L4966

If we were to implement some sort of struct equality for T::Struct, we should infinitely loop in this case which is what other typed languages like scala / kotlin / etc. do.

**Comment by ric2b** [comment]
IMO, as someone migrating a lot of (data) classes built with 
```ruby
class NewClass < Struct.new(...)
```

I would love for `T::Struct` to support more of the features of `Struct`, like equality checking and implementing `Enumerable`. It's quite laborious to implement it by hand.

**Comment by samshadwell** [comment]
I wonder, @jez would you be amenable to (someone) creating a class that sub-classes `T::Struct`, and provides some of the same utility methods as the standardlib Struct?

Names are hard, but we could start with something like `T::StdLibCompatibleStruct`. What are your thoughts? Then folks can opt-in to stdlib semantics, but nothing at Stripe would break.

**Comment by ric2b** [comment]
Regardless of whether it's compatible with the standard `Struct` equality semantics, I think `T::Struct` instances should definitely be comparable between each other. 

Re-implementing `==` for each `T::Struct` subclass is ugly boilerplate and error-prone (if someone adds an attribute but forgets to update `==`).

**Comment by michaelwebb76** [comment]
We've implemented https://github.com/tricycle/sorbet-struct-comparable to deal with this.

**Comment by MiguelSCruz** [comment]
Are there any plans to include this feature in Sorbet any time soon?

**Comment by mihyaeru21** [comment]
> Names are hard, but we could start with something like T::StdLibCompatibleStruct. What are your thoughts? Then folks can opt-in to stdlib semantics, but nothing at Stripe would break.

I agree with this opinion. With the track record of variants of `T::Struct` already implemented (#6005), it would make sense to create a new class that adds functionality to `T::Struct` with added functionality.
